core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java

```diff
@@ -71,26 +73,51 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
         int positionCount = sliceInput.readInt();
         int nonNullsCount = sliceInput.readInt();
 
+        if (nonNullsCount > positionCount) {
+            throw new IllegalArgumentException(format("nonNullsCount must be <= positionCount, found: %s > %s", nonNullsCount, positionCount));
+        }
+
         int[] offsets = new int[positionCount + 1];
-        int[] lengths = new int[nonNullsCount];
-        sliceInput.readBytes(Slices.wrappedIntArray(lengths), 0, nonNullsCount * SIZE_OF_INT);
+        // Read the lengths array into the end of the offsets array, since nonNullsCount <= positionCount
+        int lengthIndex = offsets.length - nonNullsCount;
+        sliceInput.readBytes(Slices.wrappedIntArray(offsets, lengthIndex, nonNullsCount));
 
         boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
+        // Transform lengths back to offsets
+        if (valueIsNull == null) {
+            if (positionCount != nonNullsCount || lengthIndex != 1) {
+                throw new IllegalArgumentException(format("nonNullsCount must equal positionCount, found: %s <> %s", nonNullsCount, positionCount));
+            }
+            // Simplified loop for no nulls present
+            for (int i = 1; i < offsets.length; i++) {
+                offsets[i] += offsets[i - 1];
+            }
+        }
+        else {
+            int currentOffset = 0;
+            for (int i = 1; i < offsets.length; i++) {
+                if (lengthIndex == offsets.length) {
+                    // Populate remaining null elements
+                    Arrays.fill(offsets, i, offsets.length, currentOffset);
+                    break;
+                }
+                boolean isNull = valueIsNull[i - 1];
+                // must be accessed unconditionally, otherwise CMOV optimization isn't applied due to
+                // ArrayIndexOutOfBoundsException checks
+                int length = offsets[lengthIndex];
+                lengthIndex = isNull ? lengthIndex : lengthIndex + 1;
+                currentOffset = isNull ? currentOffset : currentOffset + length;
+                offsets[i] = currentOffset;
+            }
+            if (lengthIndex != offsets.length) {
+                throw new IllegalArgumentException("Invalid nonNullsCount");
```

Comment:
Please print the values suspected to be invalid in the exception message

Reply:
Done